x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java

```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
+                    if (leading) {
+                        if (hits.length == 0) {
+                            continue;
+                        }
+                        Timestamp hitTimestamp = criterion.timestamp(hits[0]);
+                        lastLeading = lastLeading == null || lastLeading.delta(hitTimestamp) < 0 ? hitTimestamp : lastLeading;
+                    } else if (trailing(i)) {
+                        if (hits.length == 0) {
+                            continue;
+                        }
+                        Timestamp hitTimestamp = criterion.timestamp(hits[0]);
+                        firstTrailing = firstTrailing == null || firstTrailing.delta(hitTimestamp) > 0 ? hitTimestamp : firstTrailing;
+                    } else {
+                        if (hits.length > 0) {
+                            discarded = true;
+                        }
+                    }
+                } else {
+                    leading = false;
+                }
+            }
+            if (discarded == false) {
+                int lastStage = criteria.size() - 1;
+                if ((firstTrailing == null && lastLeading == null)
+                    || (lastLeading == null && matcher.isMissingEvent(0))
+                    || (firstTrailing == null && matcher.isMissingEvent(lastStage))
+                    || (matcher.isMissingEvent(0) && matcher.isMissingEvent(lastStage) && biggerThanMaxSpan(lastLeading, firstTrailing))
+                    || (matcher.isMissingEvent(0)
+                        && matcher.isMissingEvent(lastStage) == false
+                        && biggerThanMaxSpan(lastLeading, sequence.ordinal().timestamp()))
+                    || (matcher.isMissingEvent(0) == false
+                        && matcher.isMissingEvent(lastStage)
+                        && biggerThanMaxSpan(sequence.startOrdinal().timestamp(), firstTrailing))
+
+                ) {
+                    matcher.addToCompleted(sequence);
+                }
+            }
+        }
+        checkMissingEvents(next, listener);
+    }
+
+    private boolean biggerThanMaxSpan(Timestamp from, Timestamp to) {
+        if (from == null || to == null) {
+            return true;
+        }
+        return to.instant().toEpochMilli() - from.instant().toEpochMilli() >= matcher.maxSpanInNanos() / 1_000_000;
+    }
+
+    private List<SearchRequest> prepareQueryForMissingEvents(List<Sequence> toCheck) {
+        List<SearchRequest> result = new ArrayList<>();
+        for (Sequence sequence : toCheck) {
+            boolean leading = true;
+            for (int i = 0; i < criteria.size(); i++) {
+                SequenceCriterion criterion = criteria.get(i);
+                if (criterion.missing()) {
+                    BoxedQueryRequest r = criterion.queryRequest();
+                    RangeQueryBuilder range = new RangeQueryBuilder(r.timestampField()).timeZone("UTC").format("epoch_millis");
```

Comment:
Externalize this range query builder to `BoxedQueryRequest`.
```suggestion
                    RangeQueryBuilder range = RangeQueryBuilder range = r.timestampRangeQuery();
```